cli/cmd/start_test.go

```diff
@@ -0,0 +1,77 @@
+package cmd
+
+import (
+	"testing"
+
+	"github.com/spf13/cobra"
+	"github.com/stretchr/testify/assert"
+	"github.com/tarantool/tt/cli/process_utils"
+	"github.com/tarantool/tt/cli/running"
+)
+
+func TestGetStartValidArgs(t *testing.T) {
+	savedGetStatus := getStatus
+	defer func() {
+		getStatus = savedGetStatus
+	}()
+
+	instances := []running.InstanceCtx{
+		{AppName: "single_run", SingleApp: true},
+		{AppName: "app1", InstName: "master"},
+		{AppName: "app1", InstName: "replica"},
+		{AppName: "single_stopped", SingleApp: true},
+		{AppName: "app2", InstName: "master"},
+		{AppName: "app2", InstName: "s1-replica"},
+		{AppName: "app2", InstName: "s2-replica"},
+		{AppName: "app2", InstName: "s3-replica"},
+	}
+
+	statusByInstName := map[string]process_utils.ProcessState{
+		"single_run":      process_utils.ProcStateRunning,
+		"app1:master":     process_utils.ProcStateRunning,
+		"app1:replica":    process_utils.ProcStateRunning,
+		"single_stopped":  process_utils.ProcStateStopped,
+		"app2:master":     process_utils.ProcStateRunning,
+		"app2:s1-replica": process_utils.ProcStateStopped,
+		"app2:s2-replica": process_utils.ProcStateRunning,
+		"app2:s3-replica": process_utils.ProcStateDead,
+	}
+
+	getStatus = func(instCtx *running.InstanceCtx) process_utils.ProcessState {
+		return statusByInstName[running.GetAppInstanceName(*instCtx)]
+	}
+
+	cases := []struct {
+		name         string
+		instances    []running.InstanceCtx
+		toComplete   string
+		expectedArgs []string
+	}{
+		{
+			name:         "empty toComplete",
+			instances:    instances,
+			toComplete:   "",
+			expectedArgs: []string{"single_stopped", "app2"},
+		},
+		{
+			name:         "toComplete with delimiter",
+			instances:    instances,
+			toComplete:   "app:repli",
+			expectedArgs: []string{"single_stopped", "app2:s1-replica", "app2:s3-replica"},
```

Comment:
According to the line above, I'm going to complete all stopped instances starting with `repli` of an `app` application. Why does auto-completion show instances of the other apps: `single_stopped` and `app2`?

Reply:
This kind of filtering logic happens at lower levels. This function just provides the list of possible arguments, then terminal auto-completion mechanism (like `complete` in bash) do further work.